packages/container/lib/container.js

```diff
@@ -127,7 +136,96 @@ Container.prototype = {
    */
   lookupFactory(fullName, options) {
     assert('fullName must be a proper full name', this.registry.validateFullName(fullName));
-    return factoryFor(this, this.registry.normalize(fullName), options);
+
+    if (isFeatureEnabled('container-factoryFor')) {
+      deprecate('Using "_lookupFactory" is deprecated. Please use container.factoryFor instead.', false, { id: 'container-lookupFactory', until: '2.13.0', url: 'TODO' });
+    }
+
+    return deprecatedFactoryFor(this, this.registry.normalize(fullName), options);
+  },
+
+  factoryFor(fullName, options = {}) {
+    let container = this;
+    let manager;
+    if (isFeatureEnabled('container-factoryFor')) {
+      let normalizedName = this.registry.normalize(fullName);
+      assert('fullName must be a proper full name', this.registry.validateFullName(normalizedName));
+
+      if (options.source) {
+        normalizedName = this.registry.expandLocalLookup(fullName, options);
+        // if expandLocalLookup returns falsey, we do not support local lookup
+        if (!normalizedName) { return; }
+      }
+
+      let factory = this.registry.resolve(normalizedName);
+
+      if (factory === undefined) { return; }
+
+      manager = {
+        class: factory,
+        create(options = {}) {
+          let injections = injectionsFor(container, normalizedName);
+          let props = assign({}, injections, options);
+
+          runInDebug(() => {
+            let lazyInjections;
+            let validationCache = container.validationCache;
+            // Ensure that all lazy injections are valid at instantiation time
+            if (!validationCache[fullName] && factory && typeof factory._lazyInjections === 'function') {
+              lazyInjections = factory._lazyInjections();
+              lazyInjections = container.registry.normalizeInjectionsHash(lazyInjections);
+
+              container.registry.validateInjections(lazyInjections);
+            }
+
+            validationCache[fullName] = true;
+          });
+
+          this.class[NAME_KEY] = container.registry.makeToString(factory, fullName);
```

Comment:
I don't believe we want to mutate the `class`, note: it is perfectly legal to register a class multiple times with different injection rules/similar for each. When this occurs, nothing should collide and currently NAME_KEY would collide. I believe NAME_KEY can be provided merely as a member of `props`. I believe in most cases this should work IFF

https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/system/core_object.js#L546

changed to:

```js
var ret = '<' + (this[NAME_KEY] || this.constructor.toString()) + ':' + guidFor(this) + extension + '>';
```

This would mean, `class.toString()` wouldn't be terribly useful, but I don't believe it can be accomplished